java/ql/lib/semmle/code/java/environment/SystemProperty.qll

```diff
@@ -0,0 +1,123 @@
+    or
+    f.getName() = "JAVA_CLASS_VERSION" and propertyName = "java.class.version"
+    or
+    f.getName() = "JAVA_COMPILER" and propertyName = "java.compiler"
+    or
+    f.getName() = "JAVA_EXT_DIRS" and propertyName = "java.ext.dirs"
+    or
+    f.getName() = "JAVA_HOME" and propertyName = "java.home"
+    or
+    f.getName() = "JAVA_IO_TMPDIR" and propertyName = "java.io.tmpdir"
+    or
+    f.getName() = "JAVA_LIBRARY_PATH" and propertyName = "java.library.path"
+    or
+    f.getName() = "JAVA_RUNTIME_NAME" and propertyName = "java.runtime.name"
+    or
+    f.getName() = "JAVA_RUNTIME_VERSION" and propertyName = "java.runtime.version"
+    or
+    f.getName() = "JAVA_SPECIFICATION_NAME" and propertyName = "java.specification.name"
+    or
+    f.getName() = "JAVA_SPECIFICATION_VENDOR" and propertyName = "java.specification.vendor"
+    or
+    f.getName() = "JAVA_UTIL_PREFS_PREFERENCES_FACTORY" and
+    propertyName = "java.util.prefs.PreferencesFactory"
+    or
+    f.getName() = "JAVA_VENDOR" and propertyName = "java.vendor"
+    or
+    f.getName() = "JAVA_VENDOR_URL" and propertyName = "java.vendor.url"
+    or
+    f.getName() = "JAVA_VERSION" and propertyName = "java.version"
+    or
+    f.getName() = "JAVA_VM_INFO" and propertyName = "java.vm.info"
+    or
+    f.getName() = "JAVA_VM_NAME" and propertyName = "java.vm.name"
+    or
+    f.getName() = "JAVA_VM_SPECIFICATION_NAME" and propertyName = "java.vm.specification.name"
+    or
+    f.getName() = "JAVA_VM_SPECIFICATION_VENDOR" and propertyName = "java.vm.specification.vendor"
+    or
+    f.getName() = "JAVA_VM_VENDOR" and propertyName = "java.vm.vendor"
+    or
+    f.getName() = "JAVA_VM_VERSION" and propertyName = "java.vm.version"
+    or
+    f.getName() = "LINE_SEPARATOR" and propertyName = "line.separator"
+    or
+    f.getName() = "OS_ARCH" and propertyName = "os.arch"
+    or
+    f.getName() = "OS_NAME" and propertyName = "os.name"
+    or
+    f.getName() = "OS_VERSION" and propertyName = "os.version"
+    or
+    f.getName() = "PATH_SEPARATOR" and propertyName = "path.separator"
+    or
+    f.getName() = "USER_COUNTRY" and propertyName = "user.country"
+    or
+    f.getName() = "USER_DIR" and propertyName = "user.dir"
+    or
+    f.getName() = "USER_HOME" and propertyName = "user.home"
+    or
+    f.getName() = "USER_LANGUAGE" and propertyName = "user.language"
+    or
+    f.getName() = "USER_NAME" and propertyName = "user.name"
+    or
+    f.getName() = "USER_TIMEZONE" and propertyName = "user.timezone"
+  )
+}
+
+private MethodAccess getSystemPropertyFromApacheFileUtils(string propertyName) {
+  exists(Method m |
+    result.getMethod() = m and
+    m.getDeclaringType().hasQualifiedName("org.apache.commons.io", "FileUtils")
+  |
+    m.hasName(["getTempDirectory", "getTempDirectoryPath"]) and propertyName = "java.io.tmpdir"
+    or
+    m.hasName(["getUserDirectory", "getUserDirectoryPath"]) and propertyName = "user.home"
+  )
+}
```

Comment:
Both `getTempDirectory` and `getUserDirectory` return `File` instead of `String`:

 - https://github.com/apache/commons-io/blob/39b61a422e6bf87c47f15883a5b2f0d785e560fb/src/main/java/org/apache/commons/io/FileUtils.java#L1459-L1468
  - https://github.com/apache/commons-io/blob/39b61a422e6bf87c47f15883a5b2f0d785e560fb/src/main/java/org/apache/commons/io/FileUtils.java#L1481-L1490

I figure that this is still fine as it's essentially equivalent to the string variant, but wrapped in a different type

Reply:
Maybe it should be documented for `getSystemProperty` then, that the type of the expression is in some cases something other than a `String` (e.g. here a `File`, or for the separators a `char`).

Reply:
Done